cpp/include/raft/mdarray.h

```diff
@@ -0,0 +1,408 @@
+  using layout_type  = LayoutPolicy;
+  using mapping_type = typename layout_type::template mapping<extents_type>;
+
+  using index_type            = std::size_t;
+  using size_type             = std::size_t;
+  using difference_type       = std::ptrdiff_t;
+  using container_policy_type = ContainerPolicy;
+  using container_type        = typename container_policy_type::container_type;
+
+  static_assert(!std::is_const<ElementType>::value,
+                "Element type for container must not be const.");
+
+  using pointer         = typename container_policy_type::pointer;
+  using const_pointer   = typename container_policy_type::const_pointer;
+  using reference       = typename container_policy_type::reference;
+  using const_reference = typename container_policy_type::const_reference;
+
+ private:
+  template <typename E,
+            typename ViewAccessorPolicy =
+              std::conditional_t<std::is_const_v<E>,
+                                 typename container_policy_type::const_accessor_policy,
+                                 typename container_policy_type::accessor_policy>>
+  using view_type_impl =
+    std::conditional_t<container_policy_type::is_host_type::value,
+                       host_mdspan<E, extents_type, layout_type, ViewAccessorPolicy>,
+                       device_mdspan<E, extents_type, layout_type, ViewAccessorPolicy>>;
+
+ public:
+  /**
+   * \brief the mdspan type returned by view method.
+   */
+  using view_type       = view_type_impl<element_type>;
+  using const_view_type = view_type_impl<element_type const>;
+
+ public:
+  constexpr mdarray() noexcept(std::is_nothrow_default_constructible<container_type>::value) =
+    default;
+  constexpr mdarray(mdarray const&) noexcept(
+    std::is_nothrow_copy_constructible<container_type>::value) = default;
+  constexpr mdarray(mdarray&&) noexcept(std::is_nothrow_move_constructible<container_type>::value) =
+    default;
+
+  auto operator =(mdarray&&) noexcept(std::is_nothrow_move_assignable<container_type>::value)
+    -> mdarray& = default;
+
+  auto operator =(mdarray const&) noexcept(std::is_nothrow_copy_assignable<container_type>::value)
+    -> mdarray& = default;
+
+  ~mdarray() noexcept(std::is_nothrow_destructible<container_type>::value) = default;
+
+  /**
+   * @brief The only constructor that can create storage, this is to make sure CUDA stream is being
+   * used.
+   */
+  constexpr mdarray(mapping_type const& m, container_policy_type const& cp)
+    : cp_(cp), map_(m), c_(cp_.create(map_.required_span_size()))
+  {
+  }
+  /**
+   * @brief Get a mdspan that can be passed down to CUDA kernels.
+   */
+  auto view() noexcept { return view_type(c_.data(), map_, cp_.view()); }
+  /**
+   * @brief Get a mdspan that can be passed down to CUDA kernels.
+   */
+  auto view() const noexcept { return const_view_type(c_.data(), map_, cp_.view()); }
+
+  [[nodiscard]] constexpr auto size() const noexcept -> size_type { return this->view().size(); }
+
```

Comment:
`resize` functionality would be helpful. It would be tricky to promise `std::vector::resize` though. If we resize to a larger size, keeping the values at lower indices the same as the old values is not trivial.

Reply:
We need to be concerned about the `resize` method might violate the shape of the `mdarray`.

Reply:
@rg20 are you suggesting resize would allocate a larger / smaller chunk of device memory (like std::vector::resize, which changes the data shape) or are you suggesting a numpy reshape where we change the shape but the amount of memory stays the same? 

Won't changing the amount of memory allocated will always have an impact on the resulting shape?

Reply:
I am talking more of `std::vector::resize`. Consider an example where I have an `mdarray` as a member variable but I would only initialize it much later (I can do this because there is a `default` constructor). Typically, we would just do resize whenever we need (for `std::vector` or `rmm::device_uvector`).  

We can still manage with the existing implementation because we can just create a new 'mdarray' and just `move` it. My suggestion is more of making it look like other containers.

Reply:
I'm guessing we coulda accomplish this with dynamic extents?

Reply:
Ah, I did not realize that the extents in the current implementation are compile-time and so is the specification of C++ standard.  

We could only do this using dynamic extents.  Is there a particular reason (other than the standard specification) the extents have to be compile time?

Reply:
> Is there a particular reason (other than the standard specification) the extents have to be compile time?

The spec is reason enough. It is that way because there are optimization benefits to extents being statically known when possible, e.g., for loop unrolling, vectorization, modulo optimization, etc.

Reply:
Looks like the extent itself can be dynamic and is supported by the spec and this implementation. So when the extent is dynamic, we could allow resizing.